Feature 1795 layer search

yvonnesjy opened the pull request:
Add searching functionality to the layers panel- compatible with both "layers" view and "layerCategories" view. Update the view finder panel to share the search box component with the layers panel.

Fixes https://github.com/NCEAS/metacatui/issues/1795

NOTE: The base of this PR is feature-1773-layer-categories because the search implementation is an extension of the category work. I want to get the review started now. When the base branch is submitted, I'll create a new proper PR that references this one.

yvonnesjy committed:
Layer Search: Create LayersPanelView

It'll serve as a wrapper around LayerCategoryItemView OR LayerListView depending on the map config. We'll also add the search box to this view in a later commit.

yvonnesjy committed:
Layer Search: Add search support to flat layers view

The current css for coloring is arbitrary and does not follow the mock. I'll do the UI update for the layers panel in a separate PR.

yvonnesjy committed:
Layer Search: Add search support to categorized layers view

yvonnesjy committed:
Layer Search: Create SearchInputView

This replaces the current search field that's built into ViewFinderView. The component will be shared with LayersPanelView.

yvonnesjy committed:
Layer Search: Add SearchInput in LayersPanelView

yvonnesjy commented on src/js/views/maps/SearchInputView.js:
The implementation here isn't exactly the same as viewfinder, though the functionality should be noop.

@ianguerin lmk if you have any concern.

ianguerin replied:
I can't seem to find any consistent guidance on how to create shared components/widgets (whatever they're called in Backbone) using Backbone JS. I asked Robyn for an example in our codebase and am waiting to hear back.

Without having to update the shared component each time that I want to tweak the functionality of my implementation code, I'm imagining a slightly less stateful View that would just make it easy to share CSS, manage the "cancel" vs "search" button visibility. Then as a user of this shared view, I can bind my own events (I need keyup, keydown, change, etc..)

yvonnesjy replied:
Pushed a new commit for what we discussed. I left the new event handlers out for you to add when you update ViewfinderView

ianguerin commented on src/css/map-view.css:
```diff
@@ -37,6 +37,10 @@
   --map-col-yellow: #f8b406;
   --map-col-buttons-emphasis: var(--portal-secondary-color, var(--map-col-green));
   --map-col-highlight: var(--portal-primary-color, var(--map-col-blue));
+  --map-input-box-error-background: var(--portal-secondary-color-transparent);  /* #FFFCF4; */
```
I think the variable value comment is likely to become stale if the variable changes in the future

yvonnesjy replied:
oh I forgot to add a proper comment- it is the actual yellow color from the mock. I'll incorporate it in the next UI update PR.

yvonnesjy replied:
Updated

ianguerin commented on src/js/views/maps/LayerItemView.js:
```diff
@@ -486,9 +489,35 @@ define(
+        },
 
+        /**
+         * Searches and only dispays self if layer label matches the text. Highlights the
```
s/dispays/displays

yvonnesjy replied:
Done

ianguerin commented on src/js/views/maps/LayerListView.js:
```diff
@@ -119,47 +119,43 @@ define(
 
-
+        /**
+         * Searches and only dispays layers that match the text.
```
s/dispays/displays

yvonnesjy replied:
Done

ianguerin commented on src/js/views/maps/ViewfinderView.js:
Can we hold off on migrating ViewfinderView to reusable component, this is going to create some very complex merge conflicts

yvonnesjy replied:
For sure!

yvonnesjy commented:
@ianguerin Thinking through the requirements- to share CSS and cancel/search visibility, at the minimum we need a View class, sharing a template won't be sufficient. Then the customizable events will trigger from this class. Does it work if we just pass in the event handlers from parent view?

Alternatively, inheritance? https://danhough.com/blog/backbone-view-inheritance/

yvonnesjy committed:
Layer Search: Expose setError() as an API for parent views of SearchInput

Synced with @ianguerin and decided it'd be better for the parent views to have control on the error message directly.

I still included changes to ViewFinderView to show how it'd work. I'll revert the changes once SearchInput looks good to everyone.

yvonnesjy committed:
Layer Search: Comment readability fixes

yvonnesjy committed:
Layer Search: Replace native .reduce with underscore

... so that an undefined value doesn't throw an error.

yvonnesjy commented:
FYI the force push here^ is because the base branch got more commits- idt there's a way around it.
